File: parakeet_api/store.py
"""SQLite queue with fenced leases, bounded uploads and short transactions."""

import json
import sqlite3
import threading
import time
import uuid
from contextlib import contextmanager, suppress

from fastapi import HTTPException


class Store:
    def __init__(self, settings):
        self.s = settings
        self.s.data.mkdir(parents=True, exist_ok=True)
        self.blobs = self.s.data / "audio"
        self.blobs.mkdir(exist_ok=True)
        self.path = self.s.data / "jobs.sqlite3"
        if self.s.db_mode not in ("memory", "file"):
            raise ValueError("DB_MODE must be memory or file")
        self.lock = threading.RLock()
        self.memory = None
        if self.s.db_mode == "memory":
            if self.path.exists():
                raise ValueError(
                    "Use a fresh DATA_DIR for memory mode, or DB_MODE=file for existing SQLite data"
                )
            # This directory belongs exclusively to this API instance. A memory
            # queue cannot recover uploads left by its previous process.
            for path in self.blobs.iterdir():
                path.unlink()
            self.memory = sqlite3.connect(":memory:", check_same_thread=False)
        with self.connect() as db:
            db.execute("PRAGMA journal_mode=" + ("MEMORY" if self.memory is not None else "WAL"))
            db.execute("PRAGMA secure_delete=ON")
            version = db.execute("PRAGMA user_version").fetchone()[0]
            if version not in (0, 1):
                raise RuntimeError("Unsupported database schema")
            db.executescript("""
                CREATE TABLE IF NOT EXISTS uploads (
                    id TEXT PRIMARY KEY, bytes INTEGER NOT NULL,
                    ready INTEGER NOT NULL DEFAULT 0, created REAL NOT NULL);
                CREATE TABLE IF NOT EXISTS jobs (
                    id TEXT PRIMARY KEY, upload_id TEXT NOT NULL UNIQUE REFERENCES uploads(id),
                    status TEXT NOT NULL, options TEXT NOT NULL, result TEXT, error TEXT,
                    created REAL NOT NULL, updated REAL NOT NULL,
                    token TEXT, lease REAL, attempts INTEGER NOT NULL DEFAULT 0);
                CREATE INDEX IF NOT EXISTS queue ON jobs(status, created);
                PRAGMA user_version=1;
            """)

    @contextmanager
    def connect(self):
        # One long-lived connection keeps :memory: alive. Serialize transactions
        # across the HTTP event loop and cleanup thread without shared-cache mode.
        with self.lock:
            db = self.memory if self.memory is not None else sqlite3.connect(self.path, timeout=10)
            db.row_factory = sqlite3.Row
            db.execute("PRAGMA foreign_keys=ON")
            db.execute("PRAGMA secure_delete=ON")
            db.execute("PRAGMA temp_store=MEMORY")
            try:
                with db:
                    yield db
            finally:
                if self.memory is None:
                    db.close()

    def close(self):
        with self.lock:
            if self.memory is not None:
                self.memory.close()

    def reserve(self):
        with self.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            used = db.execute("SELECT coalesce(sum(bytes),0) FROM uploads").fetchone()[0]
            if used + self.s.max_bytes > self.s.storage_bytes:
                raise HTTPException(429, "Audio storage full", headers={"Retry-After": "30"})
            uid = str(uuid.uuid4())
            db.execute(
                "INSERT INTO uploads(id,bytes,created) VALUES(?,?,?)", (uid, self.s.max_bytes, time.time())
            )
            return uid

    def upload_ready(self, uid, size):
        with self.connect() as db:
            db.execute("UPDATE uploads SET ready=1,bytes=?,created=? WHERE id=?", (size, time.time(), uid))

    def discard_upload(self, uid):
        with self.connect() as db:
            db.execute(
                "DELETE FROM uploads WHERE id=? AND NOT EXISTS(SELECT 1 FROM jobs WHERE upload_id=?)",
                (uid, uid),
            )
            remaining = db.execute("SELECT 1 FROM uploads WHERE id=?", (uid,)).fetchone()
        if remaining is None:
            (self.blobs / uid).unlink(missing_ok=True)

    def submit(self, uid, options):
        with self.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            existing = db.execute("SELECT * FROM jobs WHERE upload_id=?", (uid,)).fetchone()
            if existing:
                if json.loads(existing["options"]) != options:
                    raise HTTPException(409, "Upload already submitted with different options")
                return dict(existing)
            upload = db.execute("SELECT * FROM uploads WHERE id=? AND ready=1", (uid,)).fetchone()
            if not upload:
                raise HTTPException(400, "Upload missing or expired")
            count = db.execute(
                "SELECT count(*) FROM jobs WHERE status IN ('queued','processing')"
            ).fetchone()[0]
            if count >= self.s.queue_size:
                raise HTTPException(429, "Queue full", headers={"Retry-After": "10"})
            jid, now = str(uuid.uuid4()), time.time()
            db.execute(
                "INSERT INTO jobs(id,upload_id,status,options,created,updated) VALUES(?,?,'queued',?,?,?)",
                (jid, uid, json.dumps(options), now, now),
            )
        return self.get(jid)

    def get(self, jid):
        with self.connect() as db:
            row = db.execute("SELECT * FROM jobs WHERE id=?", (jid,)).fetchone()
        if not row:
            raise HTTPException(404, "Transcript missing or expired")
        return dict(row)

    def claim(self):
        now = time.time()
        claimed = None
        with self.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            db.execute(
                """UPDATE jobs SET status=CASE WHEN attempts>=? THEN 'error' ELSE 'queued' END,
                       error=CASE WHEN attempts>=? THEN 'Worker lease expired' ELSE NULL END,
                       token=NULL,lease=NULL,updated=? WHERE status='processing' AND lease<=?""",
                (self.s.attempts, self.s.attempts, now, now),
            )
            finished = self.release_audio(db)
            row = db.execute(
                """SELECT jobs.*, uploads.bytes FROM jobs JOIN uploads ON uploads.id=jobs.upload_id
                       WHERE status='queued' ORDER BY jobs.created LIMIT 1"""
            ).fetchone()
            if row:
                token = str(uuid.uuid4())
                db.execute(
                    "UPDATE jobs SET status='processing',token=?,lease=?,attempts=attempts+1,updated=? WHERE id=?",
                    (token, now + self.s.lease, now, row["id"]),
                )
                claimed = {
                    "id": row["id"],
                    "token": token,
                    "lease_seconds": self.s.lease,
                    "bytes": row["bytes"],
                    "options": json.loads(row["options"]),
                }
        self.unlink(finished)
        return claimed

    def release_audio(self, db):
        # Finished jobs never read audio again. Their upload rows stay, so submit remains idempotent.
        ids = [
            row[0]
            for row in db.execute(
                """SELECT uploads.id FROM uploads JOIN jobs ON jobs.upload_id=uploads.id
                       WHERE jobs.status IN ('completed','error') AND uploads.bytes>0"""
            )
        ]
        db.executemany("UPDATE uploads SET bytes=0 WHERE id=?", [(uid,) for uid in ids])
        return ids

    def unlink(self, ids):
        # Only after commit: a crash in between leaves a file that cleanup removes later.
        for uid in ids:
            (self.blobs / uid).unlink(missing_ok=True)

    def fenced(self, db, jid, token):
        row = db.execute(
            "SELECT * FROM jobs WHERE id=? AND status='processing' AND token=? AND lease>?",
            (jid, token, time.time()),
        ).fetchone()
        if not row:
            raise HTTPException(409, "Lease expired or job cancelled")
        return row

    def heartbeat(self, jid, token):
        with self.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            self.fenced(db, jid, token)
            db.execute("UPDATE jobs SET lease=? WHERE id=?", (time.time() + self.s.lease, jid))

    def audio(self, jid, token):
        with self.connect() as db:
            row = self.fenced(db, jid, token)
        path = self.blobs / row["upload_id"]
        if not path.is_file():
            raise HTTPException(409, "Audio no longer available")
        return path

    def finish(self, jid, token, result=None, error=None, retry=False):
        with self.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            row = self.fenced(db, jid, token)
            if result is not None:
                status = "completed"
            else:
                status = "queued" if retry and row["attempts"] < self.s.attempts else "error"
            db.execute(
                "UPDATE jobs SET status=?,result=?,error=?,updated=?,token=NULL,lease=NULL WHERE id=?",
                (
                    status,
                    json.dumps(result) if result is not None else None,
                    None if status == "queued" else error,
                    time.time(),
                    jid,
                ),
            )
            finished = self.release_audio(db)
        self.unlink(finished)

    def delete(self, jid):
        with self.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            row = db.execute("SELECT upload_id FROM jobs WHERE id=?", (jid,)).fetchone()
            if not row:
                raise HTTPException(404, "Transcript missing or expired")
            db.execute("DELETE FROM jobs WHERE id=?", (jid,))
            db.execute("DELETE FROM uploads WHERE id=?", (row[0],))
        (self.blobs / row[0]).unlink(missing_ok=True)

    def cleanup(self):
        now = time.time()
        with self.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            db.execute(
                """UPDATE jobs SET status='error',error='Job age limit exceeded',updated=?,token=NULL,lease=NULL
                       WHERE status IN ('queued','processing') AND created<?""",
                (now, now - self.s.max_job_age),
            )
            finished = self.release_audio(db)
            expired = db.execute(
                "SELECT id,upload_id FROM jobs WHERE status IN ('completed','error') AND updated<?",
                (now - self.s.retention,),
            ).fetchall()
            for row in expired:
                db.execute("DELETE FROM jobs WHERE id=?", (row[0],))
                db.execute("DELETE FROM uploads WHERE id=?", (row[1],))
            unused = db.execute(
                """SELECT id FROM uploads WHERE created<? AND
                                NOT EXISTS(SELECT 1 FROM jobs WHERE upload_id=uploads.id)""",
                (now - max(self.s.retention, 3700),),
            ).fetchall()
            for row in unused:
                db.execute("DELETE FROM uploads WHERE id=?", (row[0],))
        self.unlink(finished + [r[1] for r in expired] + [r[0] for r in unused])
        with self.connect() as db:
            db.execute("PRAGMA wal_checkpoint(TRUNCATE)")
            # Finished uploads have bytes=0; their audio file is no longer needed.
            known = {row[0] for row in db.execute("SELECT id FROM uploads WHERE bytes>0")}
        # Recover files left between committing a deletion and unlinking audio.
        for path in self.blobs.iterdir():
            with suppress(FileNotFoundError):
                if path.name not in known and path.stat().st_mtime < now - 3700:
                    path.unlink()

    def counts(self):
        with self.connect() as db:
            return dict(db.execute("SELECT status,count(*) FROM jobs GROUP BY status").fetchall())

File: internal/store/store.go
// Package store keeps the job queue in process memory and audio blobs on disk.
//
// Nothing survives a restart: New wipes the audio directory. One mutex guards all queue state.
// Audio files are removed after the state change that releases them; a file left behind by a
// failed removal is deleted by the orphan sweep in Cleanup.
package store

import (
	"crypto/rand"
	"crypto/subtle"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/adapt2move/parakeet-api/internal/formats"
)

// Status is the lifecycle state of a job.
type Status string

const (
	StatusQueued     Status = "queued"
	StatusProcessing Status = "processing"
	StatusCompleted  Status = "completed"
	StatusError      Status = "error"
)

// staleAge bounds how long unsubmitted uploads and unknown files live. It exceeds the one hour
// limit of an upload transfer.
const staleAge = 3700 * time.Second

// minUploadCharge is the least storage an upload is charged, capped at MaxUploadBytes, so that
// tiny uploads are not nearly free.
const minUploadCharge = 64 << 10

// Config holds the limits the queue enforces.
type Config struct {
	DataDir         string // audio blobs live in DataDir/audio
	MaxUploadBytes  int64  // reserved per upload while it is written
	MaxStorageBytes int64  // sum of reserved and stored audio
	MaxPendingJobs  int    // queued plus processing jobs
	Retention       time.Duration
	MaxJobAge       time.Duration
	Lease           time.Duration
	MaxAttempts     int
	Now             func() time.Time // defaults to time.Now
}

// Error is a client-safe failure with an HTTP status.
type Error struct {
	Status     int
	Message    string
	RetryAfter int // seconds; 0 means no Retry-After header
}

func (e *Error) Error() string { return e.Message }

var (
	ErrTooLarge     = &Error{Status: http.StatusRequestEntityTooLarge, Message: "Audio too large"}
	ErrEmpty        = &Error{Status: http.StatusBadRequest, Message: "Audio is empty"}
	errStorageFull  = &Error{Status: http.StatusTooManyRequests, Message: "Audio storage full", RetryAfter: 30}
	errQueueFull    = &Error{Status: http.StatusTooManyRequests, Message: "Queue full", RetryAfter: 10}
	errUploadGone   = &Error{Status: http.StatusBadRequest, Message: "Upload missing or expired"}
	errOptions      = &Error{Status: http.StatusConflict, Message: "Upload already submitted with different options"}
	errJobGone      = &Error{Status: http.StatusNotFound, Message: "Transcript missing or expired"}
	errLease        = &Error{Status: http.StatusConflict, Message: "Lease expired or job cancelled"}
	errAudioMissing = &Error{Status: http.StatusConflict, Message: "Audio no longer available"}
)

// Job is a snapshot of a job. Result is shared and must not be modified.
type Job struct {
	ID       string
	UploadID string
	Status   Status
	Language string          // empty when unset
	Result   *formats.Result // set once completed
	Error    string          // the failure message when Status is StatusError
}

// Claim is a leased job as the worker receives it.
type Claim struct {
	ID           string            `json:"id"`
	Token        string            `json:"token"`
	LeaseSeconds int               `json:"lease_seconds"`
	Bytes        int64             `json:"bytes"`
	Options      map[string]string `json:"options"`
}

type upload struct {
	id      string
	bytes   int64 // storage charged; 0 once the job finished and its audio is gone
	size    int64
	created time.Time
	job     *job
}

type job struct {
	Job
	seq      uint64 // claim order
	upload   *upload
	updated  time.Time
	created  time.Time
	token    string
	lease    time.Time
	attempts int
}

// Store is the in-memory queue. It is safe for concurrent use.
type Store struct {
	cfg   Config
	blobs string

	mu      sync.Mutex
	uploads map[string]*upload
	jobs    map[string]*job
	counts  map[Status]int
	used    int64
	seq     uint64
}

// New prepares DataDir/audio and deletes everything in it, since an in-memory queue cannot
// recover the uploads of a previous process. The caller must hold the DataDir lock.
func New(cfg Config) (*Store, error) {
	if cfg.Now == nil {
		cfg.Now = time.Now
	}
	blobs := filepath.Join(cfg.DataDir, "audio")
	if err := os.MkdirAll(blobs, 0o700); err != nil {
		return nil, fmt.Errorf("create audio directory: %w", err)
	}
	entries, err := os.ReadDir(blobs)
	for _, entry := range entries {
		if err == nil {
			err = os.RemoveAll(filepath.Join(blobs, entry.Name()))
		}
	}
	if err != nil {
		return nil, fmt.Errorf("wipe audio directory: %w", err)
	}
	return &Store{cfg: cfg, blobs: blobs, uploads: map[string]*upload{}, jobs: map[string]*job{}, counts: map[Status]int{}}, nil
}

// Save stores src as a new upload and returns its ID. It reads at most one byte beyond
// MaxUploadBytes. Errors from src are returned unchanged; nothing of a failed upload remains.
func (s *Store) Save(src io.Reader) (string, error) {
	s.mu.Lock()
	if s.used+s.cfg.MaxUploadBytes > s.cfg.MaxStorageBytes {
		s.mu.Unlock()
		return "", errStorageFull
	}
	s.used += s.cfg.MaxUploadBytes
	s.mu.Unlock()

	id := newUUID()
	size, err := s.write(id, src)
	switch {
	case err != nil:
	case size == 0:
		err = ErrEmpty
	case size > s.cfg.MaxUploadBytes:
		err = ErrTooLarge
	}
	if err != nil {
		os.Remove(s.path(id))
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.used -= s.cfg.MaxUploadBytes
	if err != nil {
		return "", err
	}
	u := &upload{id: id, bytes: max(size, min(minUploadCharge, s.cfg.MaxUploadBytes)), size: size, created: s.cfg.Now()}
	s.used += u.bytes
	s.uploads[id] = u
	return id, nil
}

func (s *Store) write(id string, src io.Reader) (int64, error) {
	f, err := os.OpenFile(s.path(id), os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
	if err != nil {
		return 0, err
	}
	size, err := io.Copy(f, io.LimitReader(src, s.cfg.MaxUploadBytes+1))
	if closeErr := f.Close(); err == nil {
		err = closeErr
	}
	return size, err
}

// DiscardUpload drops an upload that no job references, with its file.
func (s *Store) DiscardUpload(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if u := s.uploads[id]; u != nil && u.job == nil {
		s.removeUpload(u)
		os.Remove(s.path(id))
	}
}

// Submit queues a job for an upload. Submitting the upload again with the same language returns
// the existing job.
func (s *Store) Submit(uploadID, language string) (Job, error) {
	now := s.cfg.Now()
	s.mu.Lock()
	defer s.mu.Unlock()
	u := s.uploads[uploadID]
	switch {
	case u == nil:
		return Job{}, errUploadGone
	case u.job != nil && u.job.Language != language:
		return Job{}, errOptions
	case u.job != nil:
		return u.job.Job, nil
	case s.counts[StatusQueued]+s.counts[StatusProcessing] >= s.cfg.MaxPendingJobs:
		return Job{}, errQueueFull
	}
	s.seq++
	j := &job{Job: Job{ID: newUUID(), UploadID: u.id, Language: language}, seq: s.seq, upload: u, created: now}
	u.job = j
	s.jobs[j.ID] = j
	s.setStatus(j, StatusQueued)
	return j.Job, nil
}

// Get returns a snapshot of a job.
func (s *Store) Get(id string) (Job, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if j := s.jobs[id]; j != nil {
		return j.Job, nil
	}
	return Job{}, errJobGone
}

// Claim expires stale leases and leases the oldest queued job, or returns nil.
func (s *Store) Claim() *Claim {
	now := s.cfg.Now()
	s.mu.Lock()
	defer s.mu.Unlock()
	var next *job
	for _, j := range s.jobs {
		if j.Status == StatusProcessing && !j.lease.After(now) {
			if j.attempts >= s.cfg.MaxAttempts {
				s.fail(j, "Worker lease expired", now)
				continue
			}
			s.setStatus(j, StatusQueued)
		}
		if j.Status == StatusQueued && (next == nil || j.seq < next.seq) {
			next = j
		}
	}
	if next == nil {
		return nil
	}
	s.setStatus(next, StatusProcessing)
	next.token, next.lease = newUUID(), now.Add(s.cfg.Lease)
	next.attempts++
	options := map[string]string{}
	if next.Language != "" {
		options["language_code"] = next.Language
	}
	return &Claim{ID: next.ID, Token: next.token, LeaseSeconds: int(s.cfg.Lease / time.Second), Bytes: next.upload.size, Options: options}
}

// Heartbeat extends a valid lease.
func (s *Store) Heartbeat(id, token string) error {
	now := s.cfg.Now()
	s.mu.Lock()
	defer s.mu.Unlock()
	j, err := s.leased(id, token, now)
	if err == nil {
		j.lease = now.Add(s.cfg.Lease)
	}
	return err
}

// Audio opens the audio of a leased job. The caller closes the file.
func (s *Store) Audio(id, token string) (*os.File, error) {
	s.mu.Lock()
	j, err := s.leased(id, token, s.cfg.Now())
	s.mu.Unlock()
	if err != nil {
		return nil, err
	}
	f, err := os.Open(s.path(j.UploadID))
	if err != nil {
		return nil, errAudioMissing
	}
	return f, nil
}

// Finish ends a leased attempt. A result completes the job, even with retry set. Otherwise the
// job is queued again when retry is set and attempts remain, or it fails with message.
func (s *Store) Finish(id, token string, result *formats.Result, message string, retry bool) error {
	now := s.cfg.Now()
	s.mu.Lock()
	defer s.mu.Unlock()
	j, err := s.leased(id, token, now)
	if err != nil {
		return err
	}
	switch {
	case result != nil:
		j.Result = result
		s.setStatus(j, StatusCompleted)
		s.release(j, now)
	case retry && j.attempts < s.cfg.MaxAttempts:
		s.setStatus(j, StatusQueued)
	default:
		s.fail(j, message, now)
	}
	return nil
}

// Delete removes a job, its upload and its audio. A worker holding the lease is fenced out.
func (s *Store) Delete(id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	j := s.jobs[id]
	if j == nil {
		return errJobGone
	}
	s.removeJob(j)
	os.Remove(s.path(j.UploadID))
	return nil
}

// Cleanup fails jobs older than MaxJobAge, drops finished jobs untouched for Retention and
// unsubmitted uploads older than max(Retention, 3700 s), and deletes audio files that no upload
// owns once they are 3700 s old.
func (s *Store) Cleanup() error {
	now := s.cfg.Now()
	s.mu.Lock()
	for _, j := range s.jobs {
		switch {
		case (j.Status == StatusQueued || j.Status == StatusProcessing) && now.Sub(j.created) > s.cfg.MaxJobAge:
			s.fail(j, "Job age limit exceeded", now)
		case (j.Status == StatusCompleted || j.Status == StatusError) && now.Sub(j.updated) > s.cfg.Retention:
			s.removeJob(j)
		}
	}
	known := map[string]bool{}
	for id, u := range s.uploads {
		if u.job == nil && now.Sub(u.created) > max(s.cfg.Retention, staleAge) {
			s.removeUpload(u)
			os.Remove(s.path(id))
		} else if u.bytes > 0 {
			known[id] = true
		}
	}
	s.mu.Unlock()

	entries, err := os.ReadDir(s.blobs)
	if err != nil {
		return fmt.Errorf("read audio directory: %w", err)
	}
	var errs []error
	for _, entry := range entries {
		info, err := entry.Info()
		if known[entry.Name()] || err != nil || now.Sub(info.ModTime()) <= staleAge {
			continue
		}
		if err := os.Remove(filepath.Join(s.blobs, entry.Name())); err != nil && !errors.Is(err, os.ErrNotExist) {
			errs = append(errs, errors.New("remove orphaned audio file failed"))
		}
	}
	return errors.Join(errs...)
}

// Counts returns the number of jobs per status, omitting empty statuses.
func (s *Store) Counts() map[string]int {
	s.mu.Lock()
	defer s.mu.Unlock()
	counts := map[string]int{}
	for status, n := range s.counts {
		if n > 0 {
			counts[string(status)] = n
		}
	}
	return counts
}

func (s *Store) path(uploadID string) string { return filepath.Join(s.blobs, uploadID) }

func (s *Store) leased(id, token string, now time.Time) (*job, error) {
	j := s.jobs[id]
	if j == nil || j.Status != StatusProcessing || subtle.ConstantTimeCompare([]byte(j.token), []byte(token)) != 1 || !j.lease.After(now) {
		return nil, errLease
	}
	return j, nil
}

// setStatus moves a job to status and ends any lease it held.
func (s *Store) setStatus(j *job, status Status) {
	if j.Status != "" {
		s.counts[j.Status]--
	}
	s.counts[status]++
	j.Status, j.token, j.lease = status, "", time.Time{}
}

func (s *Store) fail(j *job, message string, now time.Time) {
	j.Error = message
	s.setStatus(j, StatusError)
	s.release(j, now)
}

// release deletes the audio of a finished job. The upload stays so that re-submitting it
// returns the job.
func (s *Store) release(j *job, now time.Time) {
	j.updated = now
	s.used -= j.upload.bytes
	j.upload.bytes = 0
	os.Remove(s.path(j.UploadID))
}

func (s *Store) removeJob(j *job) {
	s.counts[j.Status]--
	delete(s.jobs, j.ID)
	s.removeUpload(j.upload)
}

func (s *Store) removeUpload(u *upload) {
	s.used -= u.bytes
	delete(s.uploads, u.id)
}

// newUUID returns a random version 4 UUID in canonical lowercase form.
func newUUID() string {
	var b [16]byte
	rand.Read(b[:])
	b[6] = b[6]&0x0f | 0x40
	b[8] = b[8]&0x3f | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
}
